<a href="https://colab.research.google.com/github/elena-gazzini/2026_Intro_Python/blob/main/EG_1_1_environment_and_data_types_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> Notebook-friendly copy of `part-I/1.1-environment-and-data-types-exercises.ipynb`, generated by `tools/make_live.py`. Edit the book notebook, not this file.

In [ ]:
# --- environment setup (generated, not part of the lesson) ---
# Colab and Kaggle do not ship every package this notebook imports.
# This is a no-op in an environment that is already set up.
import importlib.util
import subprocess
import sys

for module, package in {"pooch": "pooch"}.items():
    if importlib.util.find_spec(module) is None:
        subprocess.run([sys.executable, "-m", "pip", "install", "-q", package], check=True)

# Exercises

**ℹ️ How to use these exercises**

Attempt each from scratch in the empty code cell. Name every variable for the quantity and unit it holds, and use pathlib.Path for any file paths.

## Exercise 1: Name your measurements

Create scientifically named variables for a single observation at a river gauge: a discharge of 4.7 m³ s⁻¹, a station name of your choice, an integer count of 24 hourly samples, and a boolean flag for whether discharge exceeds 5 m³ s⁻¹. Print each value together with its type.

In [ ]:
discharge_m3_s1 = 4.7
station_name = 'station_example'
n_samples = int(24)
discharge_exceeds_5_m3_s1 = discharge_m3_s1 > 5

print("Discharge:", discharge_m3_s1, "| Type:", type(discharge_m3_s1))
print("Station name:", station_name, "| Type:", type(station_name))
print("Hourly samples:", n_samples, "| Type:", type(n_samples))
print("Discharge exceeds 5 m³/s:", discharge_exceeds_5_m3_s1, "| Type:", type(discharge_exceeds_5_m3_s1))


Discharge: 4.7 | Type: <class 'float'>
Station name: station_example | Type: <class 'str'>
Hourly samples: 24 | Type: <class 'int'>
Discharge exceeds 5 m³/s: False | Type: <class 'bool'>


## Exercise 2: Unit conversions

A temperature is recorded as 18.5 °C. Compute it in kelvin and in degrees Fahrenheit. Separately, a discharge of 4.7 m³ s⁻¹ drains a catchment of area 250 km². Express the specific discharge in mm per day (use: 1 m³ s⁻¹ spread over 1 km² for one day equals 86.4 mm). Print each result to two decimals with an f-string.

In [ ]:
## first part
temp_record_celsius = 18.5
temp_record_fahrenheit = (temp_record_celsius * 9/5) + 32
temp_record_kelvin = temp_record_celsius + 273.13

print(temp_record_celsius)
print(temp_record_fahrenheit)
print(temp_record_kelvin)

## second part
discharge_m3_s1 = 4.7 # same as exercise 1
catchment_area_km2 = 250

# specific discharge in mm per day
specific_discharge_mm_day = (discharge_m3_s1 / catchment_area_km2) * 86.4

print(f"Discharge: {discharge_m3_s1:.2f} m³/s")
print(f"Catchment area: {catchment_area_km2:.2f} km²")
print(f"Specific discharge: {specific_discharge_mm_day:.2f} mm/day")


18.5
65.3
291.63
Discharge: 4.70 m³/s
Catchment area: 250.00 km²
Specific discharge: 1.62 mm/day


## Exercise 3: Operator families and a quality flag

1.   Élément de liste
2.   Élément de liste



Three daily readings arrive from a station: `reading_1_celsius` = 18.2, `reading_2_celsius` = -300.0, and `reading_3_celsius` = None (the sensor reported nothing).

1. Build a boolean `is_plausible` that is `True` when the first reading is above absolute zero and below 60 °C.
2. Build `is_implausible` for the second reading: apply the same test as in step 1, and negate it with `not`.
3. Use `is` to check whether the third reading is missing.
4. Starting from `n_valid = 0`, use `+=` twice to count how many of the first two readings pass the step 1 test. Note that `int()` casts `True` to 1 and `False` to 0.
5. The sensor is rated for -40 °C to 60 °C, both bounds included. Use `>=` and `<=` to test whether the first reading lies in that range, and `!=` to check that the first two readings differ.

In [ ]:
# naming the values
reading_1_celsius = 18.2
reading_2_celsius = -300.0
reading_3_celsius = None

# 1. Build a boolean is_plausible that is True when the first reading is above absolute zero and below 60 °C.
is_plausible = reading_1_celsius > 0 and reading_1_celsius < 60
print(f"is_plausible: {is_plausible}")

# 2. Build is_implausible for the second reading: apply the same test as in step 1, and negate it with not.
is_implausible = not(reading_1_celsius > 0 and reading_1_celsius < 60)
# is_implausible = not reading_1_celsius > 0 and not reading_1_celsius < 60
# both same results
print(f"is_implausible: {is_implausible}")

# 3. Use is to check whether the third reading is missing.
missing = reading_3_celsius is None
print(f"missing: {missing}")

# 4. Starting from n_valid = 0, use += twice to count how many of the first two readings pass the step 1 test. Note that int() casts True to 1 and False to 0.
n_valid = 0

n_valid += int(is_plausible)
# same as n_valid += int(reading_1_celsius > 0 and reading_1_celsius < 60)
n_valid += int(reading_2_celsius > 0 and reading_2_celsius < 60)

print(f"n_valid: {n_valid}")

# 5. The sensor is rated for -40 °C to 60 °C, both bounds included. Use >= and <= to test whether the first reading lies in that range, and != to check that the first two readings differ.
sensor_range = (reading_1_celsius >= -40 and reading_1_celsius <= 60)
print(f"sensor_range: {sensor_range}")

sensor_different = reading_1_celsius != reading_2_celsius
print(f"sensor_different: {sensor_different}")


is_plausible: True
is_implausible: False
missing: True
n_valid: 1
sensor_range: True
sensor_different: True


## Exercise 4: Parse a record

A logger emits records like `"2024-01-15 , JUNGFRAUJOCH , -2.3 , degC"`. Extract the date string, the station name in title case, the numeric value as a float, and the unit as written, then print them. Then print the station name in lower case as well, the form you would use in a file name.

In [ ]:

record = "2024-01-15 , JUNGFRAUJOCH , -2.3 , degC"

# Extract each field (first we have a date, then a station, a numeric value and a unit)
date_string, station_name, numeric_value, unit = record.split(",") # slipt the record when there is a comma

# Clean the values : .strip() removes unnecessary spaces around the text.
date_string = date_string.strip()
station_name = station_name.strip().title() # .title() transforms JUNGFRAUJOCH to Jungfraujoch.
numeric_value = float(numeric_value.strip()) # float() transforms "-2.3" in a number -2.3
unit = unit.strip()

# Print the extracted values
print(f"Date: {date_string}")
print(f"Station: {station_name}")
print(f"Value: {numeric_value}")
print(f"Unit: {unit}")


# Station name in lowercase for a file name
print(f"Station filename: {station_name.lower()}") # .lower() transforms Jungfraujoch to jungfraujoch

Date: 2024-01-15
Station: Jungfraujoch
Value: -2.3
Unit: degC
Station filename: jungfraujoch


## Exercise 5: Indexing and slicing a record

Weather records often arrive in a fixed shape, where each field always occupies the same
positions. Given

```python
record = "ALO01.03.2022"
```

the first three characters are the station code and the rest is a date as `dd.mm.yyyy`.

1. Using **only indexing and slicing** (no `.split()`), extract the station code, the day, the
   month, and the four-digit year, and print them.
2. Now extract the month a second way, using `.split(".")` on the date part, and use `==` to
   confirm you get the same answer.
3. In a comment, say which of the two approaches you would trust to extract the month from a
   record where the station code can be two or four characters long, and why.

In [ ]:
record = "ALO01.03.2022"

# 1. Using only indexing and slicing
station_code = record[:3] # take only the first 3 letters ; python starts from 0
print(f"station_code: {station_code}")

day = record[3:5] # take only what's in the places 3 and 4 (not 5)
print(f"day: {day}")

month = record[6:8]
print(f"month: {month}")

year = record[9:]
# same as year = record[9:13]
print(f"year: {year}")

# 2. Extract the month using split()
date_part = record[3:]
print(f"date_part: {date_part}")

date_parts = date_part.split(".")
print(f"date_parts: {date_parts}")

month_from_split = date_parts[1] # 1 because begins at 0

print(f"Month from split: {month_from_split}")
print(f"Same month: {month == month_from_split}")

day_from_split = date_parts[0]
print(f"Day from split: {day_from_split}")
print(f"Same day: {month == day_from_split}")

# 3. I would trust split(".") if the station code can be two or four characters long,
# because fixed-position slicing depends on the station code always having the same length.

station_code: ALO
day: 01
month: 03
year: 2022
date_part: 01.03.2022
date_parts: ['01', '03', '2022']
Month from split: 03
Same month: True
Day from split: 01
Same day: False




```
# Ce texte est au format code
```

Dans les exercices 1–5, surtout :
Je crée des variables → je les utilise dans Python.

Dans l'exercice 6, apprends :
Je crée une information → je la mets dans un fichier → je peux la relire.

C'est pour ça que le mot I/O apparaît :
Input / Output, donc les échanges entre Python et l'extérieur, ici un fichier.

Le but de mettre quelque chose dans un fichier n'est pas de pouvoir utiliser une variable dans un autre chunk. Pour ça, la variable suffit.
Le fichier sert surtout à conserver des données en dehors de la mémoire de Python.

## Exercise 6: Formatting π to the precision you choose

`math.pi` holds π to about 16 significant digits, and an f-string format specifier prints it to as many decimals as you ask for.

1. Create the `_files` folder if it does not exist, as in the lecture, then write the sentence
   `"This is my first I/O exercise."` to `_files/pi_exercise.txt` with mode `"w"`.
2. Reopen the file with mode `"a"` and append three more lines: one built with an f-string that reports `math.pi` to four decimal places, and two lines of your own choosing.
3. Read the file back with `.readlines()` and print its full contents.
4. Change the format specifier to two decimals instead of four, rerun, and confirm the printed value now rounds to `3.14`.

In [ ]:
import math
import os

# 1. Create the _files folder if it does not exist
os.makedirs("_files", exist_ok=True)

# Write the first line using mode "w"
with open("_files/pi_exercise.txt", "w") as file:
    file.write("This is my first I/O exercise.\n")

# 2. Reopen the file with mode "a" and append three lines
with open("_files/pi_exercise.txt", "a") as file:
    file.write(f"Pi to four decimal places: {math.pi:.2f}\n")    # Change the format specifier to two decimals instead of four {math.pi:.4f} → {math.pi:.2f}
    file.write("Python is useful for data analysis.\n")
    file.write("This is another line in my file.\n")

# 3. Read the file back with .readlines()
with open("_files/pi_exercise.txt", "r") as file:
    contents = file.readlines()

# Print the full contents
print("".join(contents))


This is my first I/O exercise.
Pi to four decimal places: 3.14
Python is useful for data analysis.
This is another line in my file.

This is my first I/O exercise.
Pi to four decimal places: 3.14
Python is useful for data analysis.
This is another line in my file.


## Exercise 7: Round-trip through a file, with an append

Write the Jungfraujoch week from the lecture — the daily maximum temperatures -8.4, -8.9, -9.2,
-9.0, -9.5 (°C) — to `_files/jfj_week.txt`, one value per line, using `pathlib` and a `with`
block. Create the `_files` folder first if it does not exist; do not rely on Exercise 6 having
made it.

1. Write the five values with mode `"w"`, one `.write()` call per value.
2. Reopen the file with mode `"a"` and append the last two days of the week, `-7.8` and `-6.1`.
3. Read the file back with `.readlines()`, cast each of the seven lines to `float` by name, and
   compute the mean by summing the seven values by hand with `+` and dividing by `len()`. Print
   it to two decimals.

## Going deeper (optional)

### Exercise 8: Same seven values, same mean?

Adding the same floats in a different order does not have to give exactly the same result,
because each `+` rounds its result to the nearest float. This exercise uses `math.isclose` from
the going-deeper box on floating-point precision.

1. Compute the mean of the Jungfraujoch week — -8.4, -8.9, -9.2, -9.0, -9.5, -7.8, -6.1 — by
   adding the values in that order and dividing by 7.
2. Compute it again, adding the same seven values in reverse order.
3. Compare the two means with `==`, and print the result.
4. Compare the same two means with `math.isclose`, and print the result.
5. In a comment, say which of the two comparisons you would put in a script that checks whether
   two calculations of the same quantity agree, and why.

In [3]:
# 1.additionner les 7 températures dans l'ordre donné :
mean_forward = (-8.4 + -8.9 + -9.2 + -9.0 + -9.5 + -7.8 + -6.1) / 7
print(mean_forward) # mean_forward contient la moyenne obtenue dans cet ordre

# 2. idem mais à l'envers :
mean_reverse = (-6.1 + -7.8 + -9.5 + -9.0 + -9.2 + -8.9 + -8.4) / 7
print(mean_reverse)

-8.414285714285715
-8.414285714285713


Mathématiquement, c'est exactement les mêmes nombres, donc les deux résultats doivent être identiques.

Et en théorie mathématique, oui.

Mais Python travaille avec des nombres flottants (`float`), qui ne représentent pas parfaitement tous les nombres décimaux. À chaque addition, il peut donc y avoir une minuscule différence d'arrondi.

In [4]:
# 3. Avec ==
print(mean_forward == mean_reverse)

False


`==` demande « Est-ce que les deux nombres sont exactement identiques ? »

Il est possible que Python réponde : `False`, même si la différence est minuscule.

In [6]:
# 4. Avec math.isclose()
import math

print(math.isclose(mean_forward, mean_reverse))

True


`math.isclose()` pose plutôt la question : « Est-ce que les deux nombres sont suffisamment proches pour qu'on puisse considérer qu'ils représentent le même résultat ? »

Dans ce cas, il devrait répondre : `True`.

Dans un script scientifique qui vérifie si deux calculs donnent le même résultat, tu utiliserais plutôt : `math.isclose()`.

Parce que tu ne veux généralement pas que ton programme considère deux résultats comme différents simplement à cause d'une infime différence d'arrondi des `float`.

Donc l'idée à retenir est :

`==`             → exactement pareil ?

`math.isclose()`  → suffisamment proche ?

Donc, la fonction `math.isclose` est plus adaptée pour comparer des valeurs à virgule flottante calculées, car de très légères différences d'arrondi peuvent survenir.

## Exercise 9: A year of station data

The exercises so far used values typed by hand. This one uses a real record: daily maximum
temperatures for 2022 from the automated weather station at Independence Municipal Airport,
Iowa, US, station code IIB. The file has one row for each of the 365 days, but 16 of them, from
25 October to 9 November, have no temperature value. Work through the steps in order, since each
step reuses variables from the one before. The same file returns in 1.2, where you turn the whole
year into a monthly summary.

The data has three columns: the station code, the day as `dd.mm.yy`, and the daily maximum
temperature **in degrees Fahrenheit**. Converting to SI units is part of the job.

<img src="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/part-I/_static/dertobisturmjagd-heaven-3176544.jpg" alt="Storm clouds over a wheat field, with wooded hills on the horizon" width="800">

<em>Image by <a href="https://pixabay.com/users/dertobisturmjagd-7096811/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3176544">Tobias Hämmer</a> from <a href="https://pixabay.com//?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3176544">Pixabay</a></em>

In [7]:
# Pre-supplied: download the data file and cache it locally.
# You do not need to understand this cell yet — fetching data is covered in the
# reproducible-data-pipelines bonus subchapter.
import pooch

data_file = pooch.retrieve(
    url="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/data/part-I/station_iib_daily_max_temp_2022.csv",
    known_hash="sha256:034755fb289b4e157e5d029995481786a359eb25a80df62c09ee83d063013144",
    fname="station_iib_daily_max_temp_2022.csv",
    path=pooch.os_cache("mlees"),
)
print("data cached at:", data_file)

data cached at: /root/.cache/mlees/station_iib_daily_max_temp_2022.csv


**Step 1.** Open the file with a `with` block in read mode and read all its lines with
`.readlines()`.

Print how many lines the file has, then print the first three lines exactly as
they come off disk.

In [ ]:
# Your solution here
# Hint: data_file is a plain string; after from pathlib import Path, Path(data_file) is a
#       path you can .open()
# Hint: with Path(data_file).open("r", encoding="utf-8") as fhandler: ...
# Hint: .readlines() gives you a list of strings, one per line
# Hint: lines[0] is the header row, not an observation
# Hint: print lines[0], lines[1] and lines[2] one at a time

1) importer `Path`

In [9]:
from pathlib import Path

2) Transformer `data_file` en chemin
`data_file` est une chaîne de caractères contenant le chemin.


In [10]:
Path(data_file)

PosixPath('/root/.cache/mlees/station_iib_daily_max_temp_2022.csv')

Cela transforme le chemin en objet `Path`, que Python sait utiliser pour ouvrir le fichier

3. Ouvrir le fichier

In [14]:
with Path(data_file).open("r", encoding="utf-8") as fhandler :
  lines = fhandler.readlines()

- `with` → Python ouvre le fichier et s'occupe correctement de le refermer ensuite.
- `Path(data_file)` → le chemin vers le fichier.
- `.open(...)` → on ouvre le fichier.
- `"r"` → **read**, donc lecture.
- `encoding="utf-8"` →  donne un nom au fichier ouvert. Ici, le cours propose `fhandler` pour lire le contenu.

4. Ensuite, l'exercice dit : "*read all its lines with `.readlines()`*". Donc à l'intérieur du `with` mettre `lines = fhandler.readlines()`. On obtient alors une liste appelée `lines`, où chaque ligne du fichier devient un élément de la liste. Donc `lines[0]` = la première ligne, `lines[1]` = la deuxième ligne, ...

5. Compter les lignes

Pour compter les éléments de la liste :

In [16]:
len(lines)
print(len(lines))

366


*Attention* : l'exercice dit 365 jours, mais `len(lines)` devrait être 366, car il y a aussi la ligne d'en-tête (365 jours + 1 lignes d'en-tête).

6. Afficher les trois premières lignes

In [17]:
print(lines[0])
print(lines[1])
print(lines[2])

station,day,max_temp_f

IIB,01.01.22,14.7

IIB,02.01.22,8.6



**Step 2.** Look at the first record (the second line of the file). Split it on the separator
and print the resulting list. Then print the *type* of each of the three fields.

Answer in a comment: are the temperatures numbers at this point?

In [ ]:
# Your solution here
# Hint: .strip() first, to remove the trailing newline
# Hint: .split(",") turns one line into a list of fields
# Hint: type() will tell you what you are really holding

1. Prendre la première vraie ligne de données

La première ligne est `lines[0]` → l'en-tête.

Donc la première observation est :

In [18]:
lines[1]

'IIB,01.01.22,14.7\n'

`readlines()` conserve le `\n` à la fin de chaque ligne, qui est un retour à la ligne qu'on ne veut pas garder donc on l'enlève:

In [23]:
first_record = lines[1].strip()
print(first_record)

IIB,01.01.22,14.7


2. Séparer les trois champs

In [24]:
first_record = first_record.split(",")
print(first_record)

['IIB', '01.01.22', '14.7']


Donc maintenant j'ai une liste de trois éléments. Et je peux vérifier le type des trois éléments :

In [25]:
print(type(first_record[0]))
print(type(first_record[1]))
print(type(first_record[2]))

<class 'str'>
<class 'str'>
<class 'str'>


Mais même la température est encore un `str` ! Parce que `.split()` sépare le texte, mais ne transforme pas automatiquement `"31.5"` en `31.5`.

Donc pour répondre à la question "*are the temperatures numbers at this point?*" : Non, les températures sont toujours des string `str` et non des nombres.

**Step 3.** Take that same first record and turn it into three properly typed, well-named
variables: the station code as text, the day as text, and the maximum temperature as a
**float in degrees Celsius**. Print them in one line with an f-string, showing the temperature
to one decimal with its unit.

Recall that °C = (°F − 32) × 5/9.

In [ ]:
# Your solution here
# Hint: name each variable for the quantity and unit it holds, e.g. day_1 and
#       max_temp_celsius_1; step 4 needs a second set of names for the second day
# Hint: f"{value:.1f} °C" formats to one decimal

Juste pour être sûre, on repart de cette étape :

In [26]:
first_record = lines[1].strip()
first_record = first_record.split(",")
print(first_record)

['IIB', '01.01.22', '14.7']


1. Créer la variable du code station

Le premier élément est le code station :

In [28]:
station_code_1 = first_record[0]
print(station_code_1)

IIB


Il reste du texte (`str`), ce qui est ce qu'on veut.

2. Créer la variable du jour

Le deuxième élément est la date :

In [29]:
day_1 = first_record[1]
print(day_1)

01.01.22


Là aussi, on la garde comme texte (`str`), car l'exercice demande explicitement *the day as text*.

3. Convertir la température en `float`

Le troisième élément ressemble à un nombre, mais on a vu à l'étape 2 qu'il est encore un `str`.

On commence donc par le convertir :

In [31]:
max_temp_fahrenheit_1 = float(first_record[2])
print(max_temp_fahrenheit_1)
print(type(max_temp_fahrenheit_1))

14.7
<class 'float'>


Maintenant que la température n'est plus un `str` mais un `float`, on peut convertir les fahrenheit en celsius.

In [32]:
max_temp_celsius_1 = (max_temp_fahrenheit_1 - 32) * 5 / 9
print(max_temp_celsius_1)
print(type(max_temp_celsius_1))


-9.61111111111111
<class 'float'>


4. Finalement, tout afficher sur une seule ligne

In [33]:
print(f"{station_code_1}, {day_1}, {max_temp_celsius_1:.1f} °C")

IIB, 01.01.22, -9.6 °C


*rappel*: `:.f1` = affiche le nombre avec 1 chiffre après la virgule.

**Step 4.** Now do the same for the *second* record (the third line of the file), so you have
two days side by side.

1. Ask which of the two days was warmer by comparing the two raw temperature **fields**, before
   any casting.
2. Ask the same question again, this time comparing the two values **after** casting to float.
3. The two answers disagree. In a comment, explain which one is right and why the other one is
   wrong.

In [ ]:
# Your solution here
# Hint: you already have the first record's fields; do the same for lines[2]
# Hint: compare the raw strings first, then compare float() of each
# Hint: give the second day its own names, e.g. day_2 and max_temp_celsius_2

In [38]:
second_record = lines[2].strip()
second_record = second_record.split(",")

station_code_2 = second_record[0]
day_2 = second_record[1]

max_temp_fahrenheit_2 = float(second_record[2])
max_temp_celsius_2 = (max_temp_fahrenheit_2 - 32) * 5 / 9

print(f"{station_code_2}, {day_2}, {max_temp_celsius_2:.1f} °C")

IIB, 02.01.22, -13.0 °C


Maintenant, première question : comparer les strings.

Avant de faire `float()`, la température originale est dans :

`first_record[2]` et `second_record[2]`.

Donc je peux faire :

In [39]:
if first_record[2] > second_record[2]:
    print("The first day was warmer.")
else:
    print("The second day was warmer.")

The second day was warmer.


Ici, `first_record[2]` et `second_record[2]` sont des strings.

Déjà, y'a un soucis car ce n'est pas correct : la température du deuxième jour (= -13.0 °C) est plus froide que celle du premier jour (= -9.6 °C). Mais c'est le but de l'exo donc on continue.

Ensuite, comparer les températures après `float()`. j'ai déjà créé `max_temp_fahrenheit_1` et `max_temp_fahrenheit_2` qui sont des `float`. Donc:

In [40]:
if max_temp_fahrenheit_1 > max_temp_fahrenheit_2:
    print("The first day was warmer.")
else:
    print("The second day was warmer.")

The first day was warmer.


Ce qui est correct cette fois-ci.

La comparaison des températures après conversion en float est correcte, car les températures sont des nombres.

La première comparaison est incorrecte car les valeurs sont encore des strings : Python compare alors les caractères comme du texte, et non les valeurs numériques.

**Step 5.** Write your own summary file, in SI units this time.

1. Build a path to `_files/station_iib_summary.csv` with `pathlib.Path`, creating the folder if
   it does not exist.
2. In one `with` block in mode `"w"`, write a header line `station,day,max_temp_celsius` and
   then one line for each of the two days you converted, with the temperature formatted to one
   decimal.
3. Reopen the file in mode `"r"`, read it back with `.read()`, and print it.

In [ ]:
# Your solution here
# Hint: summary_path = Path("_files/station_iib_summary.csv")
# Hint: summary_path.parent.mkdir(exist_ok=True)
# Hint: build each line with an f-string, and remember the trailing \n
# Hint: f"{station_code},{day_1},{max_temp_celsius_1:.1f}\n"

1. Créer le chemin du nouveau fichier

Il faut d'abord importer Path si ce n'est pas déjà fait avec `from pathlib import Path`, mais c'est déjà fait plus haut.

Donc ensuite :

In [41]:
summary_path = Path("_files/station_iib_summary.csv")
summary_path.parent.mkdir(exist_ok=True)

`summary_path.parent` correspond au dossier `_files`.

`mkdir(exist_ok=True)` crée le dossier s'il n'existe pas déjà.

2. Ecrire le fichier
Utiliser un seul `with`, en mode `"w"`:

In [42]:
with summary_path.open("w") as f:
    f.write("station,day,max_temp_celsius\n")
    f.write(f"{station_code_1},{day_1},{max_temp_celsius_1:.1f}\n")
    f.write(f"{station_code_2},{day_2},{max_temp_celsius_2:.1f}\n")

- `f.write("station,day,max_temp_celsius\n")` → écrit le titre des colonnes.
- `f.write(f"{station_code_1},{day_1},{max_temp_celsius_1:.1f}\n")` → écrit les informations du **premier jour**.
- `f.write(f"{station_code_2},{day_2},{max_temp_celsius_2:.1f}\n")` → écrit les informations du deuxième jour.

*rappel* Le `.1f` signifie une décimale.

3. Relire le fichier

Ensuite, rouvrir le fichier en mode `"r"` :

In [43]:
with summary_path.open("r") as f:
    content = f.read()

print(content)

station,day,max_temp_celsius
IIB,01.01.22,-9.6
IIB,02.01.22,-13.0

